File: mobile-flutter/lib/core/activity_text.dart
import '../models/activity_entry.dart';

/// Renders one activity row as a plain, readable sentence for the feed.
/// Mirrors `describeActivity` in `web/src/lib/activity-text.ts` verbatim,
/// with one addition: an `ActivityAction.unknown` arm for forward
/// compatibility with a server-sent `action` value this client doesn't
/// recognize yet (the TS version's `switch` has a `default` arm for the
/// same defensive reason, even though its union type is nominally
/// exhaustive).
String describeActivity(ActivityEntry entry) {
  final who = entry.actorName.isNotEmpty ? entry.actorName : 'Someone';
  return switch (entry.action) {
    ActivityAction.createdCard =>
      '$who created card "${entry.cardTitle ?? "Untitled"}" in ${entry.toList ?? "a list"}',
    ActivityAction.moved => _describeMoved(who, entry),
    ActivityAction.deletedCard =>
      '$who deleted card "${entry.cardTitle ?? "Untitled"}" from ${entry.fromList ?? "a list"}',
    ActivityAction.createdList =>
      '$who created list "${entry.toList ?? "Untitled"}"',
    ActivityAction.renamedList =>
      '$who renamed list "${entry.fromList ?? "?"}" to "${entry.toList ?? "?"}"',
    ActivityAction.deletedList =>
      '$who deleted list "${entry.fromList ?? "Untitled"}"',
    ActivityAction.unknown => '$who did something on the board',
  };
}

String _describeMoved(String who, ActivityEntry entry) {
  final fromList = entry.fromList;
  final toList = entry.toList;
  if (fromList != null && toList != null && fromList == toList) {
    return '$who reordered "${entry.cardTitle ?? "a card"}" within $toList';
  }
  return '$who moved "${entry.cardTitle ?? "a card"}" from ${fromList ?? "?"} to ${toList ?? "?"}';
}

File: web/src/lib/activity-text.ts
import type { ActivityEntry } from "@/types/activity"

/** Renders one activity row as a plain, readable sentence for the feed. */
export function describeActivity(entry: ActivityEntry): string {
  const who = entry.actorName || "Someone"
  switch (entry.action) {
    case "created_card":
      return `${who} created card "${entry.cardTitle ?? "Untitled"}" in ${entry.toList ?? "a list"}`
    case "moved":
      if (entry.fromList && entry.toList && entry.fromList === entry.toList) {
        return `${who} reordered "${entry.cardTitle ?? "a card"}" within ${entry.toList}`
      }
      return `${who} moved "${entry.cardTitle ?? "a card"}" from ${entry.fromList ?? "?"} to ${entry.toList ?? "?"}`
    case "deleted_card":
      return `${who} deleted card "${entry.cardTitle ?? "Untitled"}" from ${entry.fromList ?? "a list"}`
    case "created_list":
      return `${who} created list "${entry.toList ?? "Untitled"}"`
    case "renamed_list":
      return `${who} renamed list "${entry.fromList ?? "?"}" to "${entry.toList ?? "?"}"`
    case "deleted_list":
      return `${who} deleted list "${entry.fromList ?? "Untitled"}"`
    default:
      return `${who} did something on the board`
  }
}
